app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,474 @@
+# == Schema Information
+#
+# Table name: services
+#
+#  id                    :integer          not null, primary key
+#  type                  :string(255)
+#  title                 :string(255)
+#  project_id            :integer
+#  created_at            :datetime
+#  updated_at            :datetime
+#  active                :boolean          default(FALSE), not null
+#  properties            :text
+#  template              :boolean          default(FALSE)
+#  push_events           :boolean          default(TRUE)
+#  issues_events         :boolean          default(TRUE)
+#  merge_requests_events :boolean          default(TRUE)
+#  tag_push_events       :boolean          default(TRUE)
+#  note_events           :boolean          default(TRUE), not null
+#
+
+require 'digest'
+require 'composer'
+
+class ComposerService < Service
+  prop_accessor :package_mode, :package_type, :export_branches, :branch_filters,
+                :export_tags, :tag_filters, :custom_json
+
+  validates :package_mode, :package_type, presence: true, if: :activated?
+
+  validates_each :custom_json,
+    if: :allow_custom_json_validation? do |record, attr, value|
+      begin
+        name_re = /([A-Za-z0-9&_-]+\/[A-Za-z0-9&_-]+)/
+
+        if value.empty?
+          error = 'must be specified'
+        elsif (cjson = ActiveSupport::JSON.decode(value))
+          if cjson.empty?
+            error = 'must not be empty'
+          elsif !cjson['name']
+            error = 'must have a name key specified'
+          elsif cjson['name'].empty?
+            error = 'name key must not be empty'
+          elsif cjson['name'] != (name_re.match(cjson['name']) || [])[0]
+            error = 'name key must be formatted as "namespace/project"'
+          elsif !cjson['description']
+            error = 'must have a description key specified'
+          elsif cjson['description'].empty?
+            error = 'description key must not be empty'
+          end
+        end
+        record.errors.add(attr, error) if error
+
+      rescue
+        record.errors.add(attr, 'must be a valid JSON string')
+      end
+    end
+
+  after_save :process_project
+
+  def allow_custom_json_validation?
+    activated? && package_mode == 'advanced'
+  end
+
+  def title
+    'Composer'
+  end
+
+  def description
+    'List your project as a composer package'
+  end
+
+  def help
+    out = 'This project will be publicly listed as a composer package,
```

Comment:
@DouweM I think the work being done there is to allow composer to access a gitlab repo directly using Oauth and has also been pending for a while. (I hope it gets merged eventually)

This MR is enabling Gitlab to act as a packagist repository. I admit that there are many ways to tackle this. For example one can setup satis and use that configuration to serve up packages.

For smaller shops, like mine, doing so would be an overkill since we have a small amount of packages and having this kind of gitlab service eliminates the need to use a dedicated solution while only having to maintain our gitlab server.

I agree that this solution has room for improvement and would be more than happy to improve on it. Gitlab services have been engineered mainly under a "remote" point of view and lack in "local" perspective and are perhaps in need of some redesign. 

The renaming/changing project attributes issue affects not only this service but even Gitlab CI. What happens if you rename/move a project that is linked to Ci. From my experience I can tell you that the Ui links pointing back from CI -> Gitlab don't get updated and 404. I had to manually query the ci database and change the project paths in order to fix them. So events for these kinds of changes should be implemented so that services can sync and propagate the changes.

Let me know what you think.